Mathlib/Data/Finsupp/MonomialOrder.lean

```diff
@@ -0,0 +1,411 @@
+
+/-- The lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.lex [WellFoundedGT σ] :
+    MonomialOrder σ where
+  syn := Lex (σ →₀ ℕ)
+  locacm := Lex.linearOrderedCancelAddCommMonoid
+  toSyn := {
+    toEquiv := toLex
+    map_add' := toLex_add } -- AddEquiv.refl _ -- (AddEquiv.refl (Lex (σ →₀ ℕ))).symm
+  wf := Lex.wellFoundedLT
+  toSyn_monotone := Finsupp.toLex_monotone
+
+theorem MonomialOrder.lex_le_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≼[lex] d ↔ toLex c ≤ toLex d := Iff.rfl
+
+theorem MonomialOrder.lex_lt_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≺[lex] d ↔ toLex c < toLex d := Iff.rfl
+
+/-- The reverse lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.revLex [WellFoundedLT σ] :
+    MonomialOrder σ :=
+  MonomialOrder.lex (σ := σᵒᵈ)
+
+theorem MonomialOrder.revLex_le_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≼[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) ≤ (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+theorem MonomialOrder.revLlex_lt_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≺[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) < (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+end Lex
+
+section degLex
+
+/-- A type synonym to equip a type with its lexicographic order sorted by degrees. -/
+def DegLex (α : Type*) := α
+
+variable {α : Type*}
+/-- `toDegLex` is the identity function to the `DegLex` of a type.  -/
+@[match_pattern]
+def toDegLex : α ≃ DegLex α :=
+  Equiv.refl _
+
+/-- `ofDegLex` is the identity function from the `DegLex` of a type.  -/
+@[match_pattern]
+def ofDegLex : DegLex α ≃ α :=
+  Equiv.refl _
+
+@[simp]
+theorem toDegLex_symm_eq : (@toDegLex α).symm = ofDegLex :=
+  rfl
+
+@[simp]
+theorem ofDegLex_symm_eq : (@ofDegLex α).symm = toDegLex :=
+  rfl
+
+@[simp]
+theorem toDegLex_ofDegLex (a : DegLex α) : toDegLex (ofDegLex a) = a :=
+  rfl
+
+@[simp]
+theorem ofDegLex_toDegLex (a : α) : ofDegLex (toDegLex a) = a :=
+  rfl
+
+theorem toDegLex_inj {a b : α} : toDegLex a = toDegLex b ↔ a = b :=
+  Iff.rfl
+
+theorem ofDegLex_inj {a b : DegLex α} : ofDegLex a = ofDegLex b ↔ a = b :=
+  Iff.rfl
+
+/-- A recursor for `DegLex`. Use as `induction x`. -/
+@[elab_as_elim, induction_eliminator, cases_eliminator]
+protected def DegLex.rec {β : DegLex α → Sort*} (h : ∀ a, β (toDegLex a)) :
+    ∀ a, β a := fun a => h (ofDegLex a)
+
+@[simp] lemma DegLex.forall {p : DegLex α → Prop} : (∀ a, p a) ↔ ∀ a, p (toDegLex a) := Iff.rfl
+@[simp] lemma DegLex.exists {p : DegLex α → Prop} : (∃ a, p a) ↔ ∃ a, p (toDegLex a) := Iff.rfl
+
+noncomputable instance [AddCommMonoid α] :
+    AddCommMonoid (DegLex α) := ofDegLex.addCommMonoid
+
+theorem toDegLex_add [AddCommMonoid α] (a b : α) :
+    toDegLex (a + b) = toDegLex a + toDegLex b := rfl
+
+theorem ofDegLex_add [AddCommMonoid α] (a b : DegLex α) :
+    ofDegLex (a + b) = ofDegLex a + ofDegLex b := rfl
+
+namespace Finsupp
+
+variable {M N : Type*} [AddCommMonoid M] [CanonicallyOrderedAddCommMonoid N]
```

Comment:
These variables are not unused. In the definition of `DegLex`, the docstring refers to `M`, but the definition uses `ℕ`. `Finsupp.degree` is defined only for `ℕ`. Did you intend to get more general definitions?

Reply:
Probably,  indeed, especially because I'm not satisfied with `Finsupp.degree`, which is uselessly restricted to `ℕ`, and about which I have conflicted views.

Reply:
Meanwhile, I'll remove unused variables.